tsl/src/chunk.c

```diff
@@ -507,3 +508,187 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
 
 	PG_RETURN_VOID();
 }
+
+/*
+ * chunk_drop_stale_chunks:
+ *
+ * This function drops chunks on a specified data node if those chunks are
+ * not known by the access node (chunks array).
+ *
+ * This function is intended to be used on the access node and data node.
+ */
+Datum
+chunk_drop_stale_chunks(PG_FUNCTION_ARGS)
+{
+	const Oid table_relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	char *node_name = PG_ARGISNULL(1) ? NULL : NameStr(*PG_GETARG_NAME(1));
+	ArrayType *chunks_arr = PG_ARGISNULL(2) ? NULL : PG_GETARG_ARRAYTYPE_P(2);
+	Cache *hcache;
+	Hypertable *ht;
+	DistUtilMembershipStatus membership;
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	if (!OidIsValid(table_relid))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid hypertable: cannot be NULL")));
+
+	hcache = ts_hypertable_cache_pin();
+	ht = ts_hypertable_cache_get_entry(hcache, table_relid, CACHE_FLAG_NONE);
+
+	ts_hypertable_permissions_check(ht->main_table_relid, GetUserId());
+
+	/* execute according to the node membership */
+	membership = dist_util_membership();
+	if (membership == DIST_MEMBER_ACCESS_NODE)
+	{
+		ListCell *lc;
+		List *chunk_id_list;
+		StringInfo cmd = makeStringInfo();
+		bool first = true;
+
+		if (!hypertable_is_distributed(ht))
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_HYPERTABLE_NOT_DISTRIBUTED),
+					 errmsg("hypertable \"%s\" is not distributed", get_rel_name(table_relid))));
+
+		if (node_name == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("node_name argument cannot be NULL")));
+		if (chunks_arr != NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("chunks argument cannot be used on the access node")));
+
+		/* generate query to execute drop_stle_chunks() on the data node */
+		appendStringInfo(cmd,
+						 "SELECT _timescaledb_internal.drop_stale_chunks(%s, NULL, ",
+						 quote_literal_cstr(
+							 quote_qualified_identifier(NameStr(ht->fd.schema_name),
+														NameStr(ht->fd.table_name))));
+
+		/* serialize chunk creation around a lock on the "main table" to avoid
+		 * multiple processes trying to create new chunks */
+		LockRelationOid(ht->main_table_relid, ShareUpdateExclusiveLock);
+		ts_hypertable_has_chunks(table_relid, AccessShareLock);
+
+		/* collect a list of chunks which should exists on the specified data node
+		 * according to the access node */
+		chunk_id_list = ts_chunk_get_chunk_ids_by_hypertable_id(ht->fd.id);
```

Comment:
Instead of doing the all of these scans, can't we just scan _only_ `chunk_data_node` with a node_name index scan? 

If you want the chunks for just a specific hypertable (although I don't know why we do that), you can use ` `ts_chunk_get_relid` to map the chunk_id to a relid.